src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -329,6 +329,20 @@ public GenericDictionaryLookup ComputeGenericLookup(MethodDesc contextMethod, Re
             return GenericDictionaryLookup.CreateHelperLookup(contextSource);
         }
 
+        /// <summary>
+        /// Gets the type of System.Type descendant that implements runtime types.
+        /// </summary>
+        public virtual TypeDesc GetTypeOfRuntimeType()
+        {
+            ModuleDesc reflectionCoreModule = TypeSystemContext.GetModuleForSimpleName("System.Private.Reflection.Core", false);
```

Comment:
It would be nice to avoid baking in dependency on System.Private.Reflection.Core here.

Would it make sense to change IRuntimeImplementedType to base class, make the runtime types inherit from it, and return that base class here?

Reply:
Yes, it would be nice to not hardcode, but if we add a middle man class and report that, we lose a bunch of devirtualization that is possible on `RuntimeTypeInfo` (the class itself is not sealed, but it has a lot of sealed virtual methods). Would it make sense to try move `RuntimeTypeInfo` to CoreLib (or parts of it) instead?

Also, I was thinking that a bunch of these places where the compiler hardcodes how the class library looks like should go to a "CoreLib binder" class instead, that the compiler driver instantiates based on an XML configuration manifest placed into S.P.CoreLib. That could be another way to avoid hardcoding this in the compiler...

Reply:
I see - we can hardcoded it for now.

I think that the right eventual factoring is to have a sealed `RuntimeType` in CoreLib and make it point to cache with the heavy reflection stuff via weak handle. Just like it is in CoreCLR.

The current reflection factoring is artifact of how stuff evolved, but it has less than ideal properties:
- Even a simple `typeof(Foo)` will pull in the whole reflection stack and hit metadata. It is not pay-for-play. All that `typeof(Foo)` really needs is a hashtable that maps `EEType` to `RuntimeType` object, no need to hit metadata for it.
- The `RuntimeType` descendants hierarchy hold on too much stuff forever. It is based on assumption that people are not holding onto types (https://github.com/dotnet/corefx/blob/cea2a3e00c9a508fbf92bcaf0aeb4df833514057/src/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/RuntimeTypeInfo.TypeComponentsCache.cs#L17). I do not think it is a safe assumption to make for typical programs - the weak handle cache was introduced in CoreCLR after experimenting with different options on real workloads.